File: scripts/ci/coverage-gate.py
#!/usr/bin/env python3
"""Production-line coverage gate (T5.7).

`cargo llvm-cov` instruments every line of `src/`, including the inline
`#[cfg(test)] mod tests` blocks that hold this crate's unit tests and the
test doubles that are compiled unconditionally for the integration
suites (`src/kernel/fake.rs`); test code is nearly fully covered by
construction. This script drops the inline test blocks and the excluded
files from an LCOV report and reports the coverage of the remaining
**production lines**, which is the number the floor applies to.

    scripts/ci/coverage-gate.py LCOV [--floor PERCENT] [--exclude PATH]...
                               [--json OUT] [--markdown OUT] [--filtered OUT]

A file's inline test block is the top-level `#[cfg(test)]` attribute
immediately followed by `mod tests {`; it must run to the end of the file
(the layout every module here uses), otherwise the script fails rather
than guess. `--exclude` names a file (relative to `--root`) whose lines
count as test support: it is left out of the production figure and of
the filtered report, and still counted under "all instrumented lines".

The filtered report (`--filtered`) is a line and branch report: `DA` and
`BRDA` records carry a source line and are cut at the test block, while
function records (`FN`, `FNDA`, whose first field is an execution count,
not a line, and the `FNF`/`FNH`/`FNL`/`FNA` totals) are dropped
altogether rather than filtered by the wrong field. The `LF`/`LH` and
`BRF`/`BRH` totals are recomputed from the retained records (a branch
counts as hit when its taken count is a number above zero).

Exit status 1 when the production percentage is below the floor, 2 on a
layout or input problem.
"""
import argparse
import json
import re
import sys
from pathlib import Path

MARKER = re.compile(r"^#\[cfg\(test\)\]\s*$")


def test_block_start(path: Path) -> int | None:
    """1-based line of the `#[cfg(test)]` that opens the inline test module."""
    try:
        lines = path.read_text(encoding="utf-8").split("\n")
    except OSError:
        return None
    for i, line in enumerate(lines):
        if MARKER.match(line) and i + 1 < len(lines) and lines[i + 1].startswith("mod tests"):
            # The module must be the last top-level item: after `mod tests {`
            # every non-empty line is indented, except the closing brace.
            for j, rest in enumerate(lines[i + 2 :], start=i + 3):
                if rest and not rest[0].isspace() and rest.rstrip() != "}":
                    print(
                        f"coverage-gate: {path}:{j}: top-level code after the inline test module",
                        file=sys.stderr,
                    )
                    sys.exit(2)
            return i + 1
    return None


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("lcov")
    ap.add_argument("--floor", type=float, default=None)
    ap.add_argument("--json")
    ap.add_argument("--markdown")
    ap.add_argument("--filtered", help="write the LCOV without inline test lines")
    ap.add_argument("--root", default=".", help="repository root the LCOV paths are relative to")
    ap.add_argument(
        "--exclude",
        action="append",
        default=[],
        metavar="PATH",
        help="test-support file (relative to --root) left out of the production figure",
    )
    args = ap.parse_args()

    root = Path(args.root).resolve()
    excluded = {(root / p).resolve() for p in args.exclude}
    excluded_seen = []
    all_found = all_hit = prod_found = prod_hit = 0
    per_file = []
    out_lines = []
    current = None
    cutoff = None
    skip_file = False
    file_found = file_hit = 0
    br_found = br_hit = 0
    for raw in Path(args.lcov).read_text(encoding="utf-8").splitlines():
        if raw.startswith("SF:"):
            p = Path(raw[3:])
            current = p if p.is_absolute() else root / p
            rel = current.relative_to(root) if str(current).startswith(str(root)) else current
            skip_file = current.resolve() in excluded
            if skip_file:
                excluded_seen.append(str(rel))
            cutoff = None if skip_file else test_block_start(current)
            file_found = file_hit = 0
            br_found = br_hit = 0
            if not skip_file:
                out_lines.append(raw)
            continue
        if raw.startswith("DA:"):
            line, hits = raw[3:].split(",")[:2]
            line = int(line)
            hit = int(hits) > 0
            all_found += 1
            all_hit += hit
            if skip_file or (cutoff is not None and line >= cutoff):
                continue
            prod_found += 1
            prod_hit += hit
            file_found += 1
            file_hit += hit
            out_lines.append(raw)
            continue
        if raw.startswith("BRDA:"):
            # BRDA:<line>,<block>,<branch>,<taken>: cut at the test block.
            fields = raw[5:].split(",")
            line = int(fields[0])
            if skip_file or (cutoff is not None and line >= cutoff):
                continue
            br_found += 1
            br_hit += fields[3].isdigit() and int(fields[3]) > 0
            out_lines.append(raw)
            continue
        if raw.startswith(("FN:", "FNDA:", "FNF:", "FNH:", "FNL:", "FNA:")):
            continue  # function records: not a line report, see the module docstring
        if raw.startswith(("LF:", "LH:", "BRF:", "BRH:")):
            continue  # recomputed below
        if raw == "end_of_record":
            if current is not None and not skip_file:
                per_file.append((str(rel), file_found, file_hit))
                if br_found:
                    out_lines.append(f"BRF:{br_found}")
                    out_lines.append(f"BRH:{br_hit}")
                out_lines.append(f"LF:{file_found}")
                out_lines.append(f"LH:{file_hit}")
                out_lines.append(raw)
            current = None
            skip_file = False
            continue
        if not skip_file:
            out_lines.append(raw)

    if all_found == 0:
        print("coverage-gate: no line records in the LCOV input", file=sys.stderr)
        sys.exit(2)
    pct = lambda h, f: (100.0 * h / f) if f else 0.0
    all_pct = pct(all_hit, all_found)
    prod_pct = pct(prod_hit, prod_found)
    summary = {
        "all_lines": {"found": all_found, "hit": all_hit, "percent": round(all_pct, 2)},
        "production_lines": {"found": prod_found, "hit": prod_hit, "percent": round(prod_pct, 2)},
        "floor_percent": args.floor,
        "passed": args.floor is None or prod_pct >= args.floor,
        "excluded": sorted(excluded_seen),
    }
    md = [
        "| Scope | Lines | Covered | Coverage |",
        "|---|---|---|---|",
        f"| Production lines (inline `mod tests` and test doubles excluded) | {prod_found} | {prod_hit} | **{prod_pct:.1f}%** |",
        f"| All instrumented lines | {all_found} | {all_hit} | {all_pct:.1f}% |",
    ]
    if args.floor is not None:
        md.append(f"| Floor (production lines) | | | {args.floor:g}% |")
    md.append("")
    if excluded_seen:
        md.append("Test-support files excluded from the production figure: " + ", ".join(f"`{p}`" for p in sorted(excluded_seen)))
        md.append("")
    md.append("<details><summary>Per file (production lines)</summary>")
    md.append("")
    md.append("| File | Lines | Coverage |")
    md.append("|---|---|---|")
    for name, found, hit in sorted(per_file):
        md.append(f"| `{name}` | {found} | {pct(hit, found):.1f}% |")
    md.append("")
    md.append("</details>")
    text = "\n".join(md) + "\n"
    print(text)
    if args.markdown:
        Path(args.markdown).write_text(text, encoding="utf-8")
    if args.json:
        Path(args.json).write_text(json.dumps(summary, indent=2) + "\n", encoding="utf-8")
    if args.filtered:
        Path(args.filtered).write_text("\n".join(out_lines) + "\n", encoding="utf-8")
    if args.floor is not None and prod_pct < args.floor:
        print(f"coverage-gate: production-line coverage {prod_pct:.2f}% is below the floor {args.floor:g}%", file=sys.stderr)
        return 1
    print(f"coverage-gate: production-line coverage {prod_pct:.2f}%" + (f" (floor {args.floor:g}%)" if args.floor is not None else ""), file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/coverage_gate.rs
//! `scripts/ci/coverage-gate.py` (T5.7): the floor applies to production
//! lines only, so the inline `#[cfg(test)] mod tests` block of each file is
//! removed from the LCOV before the percentage is computed.
#![forbid(unsafe_code)]
#![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]

use std::path::Path;
use std::process::Command;

const SOURCE: &str = "pub fn a() -> u32 {\n    1\n}\n\npub fn b() -> u32 {\n    2\n}\n\n#[cfg(test)]\nmod tests {\n    #[test]\n    fn t() {\n        assert_eq!(super::a(), 1);\n    }\n}\n";

fn run(lcov: &str, source: &str, floor: &str) -> (i32, String, String) {
    let dir = tempfile::tempdir().unwrap();
    std::fs::create_dir(dir.path().join("src")).unwrap();
    std::fs::write(dir.path().join("src/m.rs"), source).unwrap();
    std::fs::write(dir.path().join("cov.info"), lcov).unwrap();
    let script = Path::new(env!("CARGO_MANIFEST_DIR")).join("scripts/ci/coverage-gate.py");
    let out = Command::new("python3")
        .arg(script)
        .arg("cov.info")
        .args([
            "--floor",
            floor,
            "--json",
            "s.json",
            "--filtered",
            "prod.info",
        ])
        .current_dir(dir.path())
        .output()
        .unwrap();
    let json = std::fs::read_to_string(dir.path().join("s.json")).unwrap_or_default();
    let filtered = std::fs::read_to_string(dir.path().join("prod.info")).unwrap_or_default();
    (
        out.status.code().unwrap(),
        format!(
            "{}{}\n{json}",
            String::from_utf8_lossy(&out.stdout),
            String::from_utf8_lossy(&out.stderr)
        ),
        filtered,
    )
}

// Production lines 1, 2, 5, 6 (a covered, b not); test lines 11-14 covered.
const LCOV: &str = "SF:src/m.rs\nDA:1,1\nDA:2,1\nDA:5,0\nDA:6,0\nDA:11,1\nDA:12,1\nDA:13,1\nDA:14,1\nLF:8\nLH:6\nend_of_record\n";

#[test]
fn inline_test_lines_are_excluded_from_the_production_figure() {
    let (code, text, filtered) = run(LCOV, SOURCE, "40");
    assert_eq!(code, 0, "{text}");
    assert!(
        text.contains(
            "\"production_lines\": {\n    \"found\": 4,\n    \"hit\": 2,\n    \"percent\": 50.0"
        ),
        "{text}"
    );
    assert!(
        text.contains(
            "\"all_lines\": {\n    \"found\": 8,\n    \"hit\": 6,\n    \"percent\": 75.0"
        ),
        "{text}"
    );
    assert!(!filtered.contains("DA:11,"), "{filtered}");
    assert!(filtered.contains("DA:5,0"), "{filtered}");
    assert!(filtered.contains("LF:4\nLH:2"), "{filtered}");
}

#[test]
fn the_floor_applies_to_production_lines_not_all_lines() {
    // All lines are at 75 %, production at 50 %: a 60 % floor fails.
    let (code, text, _) = run(LCOV, SOURCE, "60");
    assert_eq!(code, 1, "{text}");
    assert!(text.contains("below the floor 60%"), "{text}");
    assert!(text.contains("\"passed\": false"), "{text}");
}

#[test]
fn a_file_without_inline_tests_is_counted_whole() {
    let source = "pub fn a() -> u32 {\n    1\n}\n";
    let lcov = "SF:src/m.rs\nDA:1,1\nDA:2,1\nend_of_record\n";
    let (code, text, _) = run(lcov, source, "100");
    assert_eq!(code, 0, "{text}");
    assert!(text.contains("\"percent\": 100.0"), "{text}");
}

#[test]
fn production_code_after_the_test_module_is_rejected() {
    let source = format!("{SOURCE}\npub fn c() {{}}\n");
    let (code, text, _) = run(LCOV, &source, "0");
    assert_eq!(code, 2, "{text}");
    assert!(
        text.contains("top-level code after the inline test module"),
        "{text}"
    );
}

// Function records: `FNDA:<count>,<name>` carries an execution count, not a
// line, so it cannot be filtered by line; the filtered file is a line and
// branch report and carries no function records at all.
const LCOV_WITH_FUNCTIONS: &str = "SF:src/m.rs\nFN:1,a\nFN:5,b\nFN:11,tests::t\nFNDA:500,a\nFNDA:0,b\nFNDA:1,tests::t\nFNF:3\nFNH:2\nDA:1,500\nDA:2,500\nDA:5,0\nDA:6,0\nDA:11,1\nDA:12,1\nDA:13,1\nDA:14,1\nBRDA:2,0,0,500\nBRDA:12,0,0,1\nBRF:2\nBRH:2\nLF:8\nLH:6\nend_of_record\n";

#[test]
fn function_records_are_not_filtered_by_their_execution_count() {
    let (code, text, filtered) = run(LCOV_WITH_FUNCTIONS, SOURCE, "40");
    assert_eq!(code, 0, "{text}");
    for record in ["FN:", "FNDA:", "FNF:", "FNH:", "FNL:", "FNA:"] {
        assert!(
            !filtered.contains(record),
            "{record} must not survive into the line report:\n{filtered}"
        );
    }
    // Branch records do carry a line and follow the same cut as DA.
    assert!(filtered.contains("BRDA:2,0,0,500"), "{filtered}");
    assert!(!filtered.contains("BRDA:12,"), "{filtered}");
    assert!(text.contains("\"percent\": 50.0"), "{text}");
}

#[test]
fn excluded_test_support_files_do_not_count_as_production() {
    // Two files: the production module and a test double compiled
    // unconditionally. The double is fully covered and would lift the
    // production figure; `--exclude` removes it from that figure and from
    // the filtered report while the all-lines figure still counts it.
    let dir = tempfile::tempdir().unwrap();
    std::fs::create_dir_all(dir.path().join("src/kernel")).unwrap();
    std::fs::write(dir.path().join("src/m.rs"), SOURCE).unwrap();
    std::fs::write(
        dir.path().join("src/kernel/fake.rs"),
        "pub fn f() -> u32 {\n    3\n}\n",
    )
    .unwrap();
    let lcov = format!("{LCOV}SF:src/kernel/fake.rs\nDA:1,1\nDA:2,1\nend_of_record\n");
    std::fs::write(dir.path().join("cov.info"), lcov).unwrap();
    let script = Path::new(env!("CARGO_MANIFEST_DIR")).join("scripts/ci/coverage-gate.py");
    let out = Command::new("python3")
        .arg(script)
        .args([
            "cov.info",
            "--floor",
            "55",
            "--exclude",
            "src/kernel/fake.rs",
            "--json",
            "s.json",
            "--filtered",
            "prod.info",
        ])
        .current_dir(dir.path())
        .output()
        .unwrap();
    let text = format!(
        "{}{}\n{}",
        String::from_utf8_lossy(&out.stdout),
        String::from_utf8_lossy(&out.stderr),
        std::fs::read_to_string(dir.path().join("s.json")).unwrap_or_default()
    );
    // With the double counted the production figure would be 4/6 = 66.7 %
    // and pass a 55 % floor; without it, 2/4 = 50 % fails.
    assert_eq!(out.status.code().unwrap(), 1, "{text}");
    assert!(
        text.contains("\"found\": 4,\n    \"hit\": 2,\n    \"percent\": 50.0"),
        "{text}"
    );
    assert!(
        text.contains("\"found\": 10,\n    \"hit\": 8,\n    \"percent\": 80.0"),
        "{text}"
    );
    assert!(
        text.contains("\"excluded\": [\n    \"src/kernel/fake.rs\"\n  ]"),
        "{text}"
    );
    let filtered = std::fs::read_to_string(dir.path().join("prod.info")).unwrap();
    assert!(!filtered.contains("SF:src/kernel/fake.rs"), "{filtered}");
    assert!(filtered.contains("SF:src/m.rs"), "{filtered}");
}
